autoprecompiles/src/optimizer_documentation.md

```diff
@@ -0,0 +1,221 @@
+  with the given ID is stateful or not. If a bus interaction is not stateful,
+  it only affects the payload passed to it and no other elements of the system.
+  A memory bus or the execution bridge are examples of a stateful busses,
+  while range constraint busses, busses modeling arithmetic operations or
+  lookup tables are not stateful.
+
+- `handle_bus_interaction`: Takes a Bus Interaction where its items are represented
+  by Range Constraints instead of expressions or field elements and returns
+  an equivalent Bus Interaction where the Range Constraints are possibly
+  tightened. A correct implementation would be one that always returns
+  its input (or also just fully unconstrained Range Constraints), but the tighter
+  the Range Constraints are, the more powerful the optimizer will be.
+
+As an example, let us assume we are modeling a bus that implements a byte range
+constraint, i.e. a bus that takes a single payload item and enforces that it is
+in the range `0..=255`. The bus is not stateful since it does not depend on any
+parts of the system. A simple (and also best possible) implementation of
+`handle_bus_interaction` would be to always return the `0xff`-mask Range Constraint
+for the payload and ignore the input. Even if the input Range Constraint is something
+like `200..=300`, the solver will combine the two Range Constraints and derive
+`200..=255` as the new Range Constraint for the payload.
+
+Another example would be an XOR-bus that takes three payload items `a, b, c`
+and ensures that all of them are bytes and `a ^ b = c`. This bus is also not stateful.
+Here, one would implement `handle_bus_interaction` by returning the three byte constraints
+for the payload items if the input has no restrictiens. But if the value of the same bit
+in two inputs is known, we can compute the value of the same bit in the third input.
+Since our Range Constraints do not model individual bits well enough, it is sufficient
+to only compute the third item if two of them are fully known.
+
+We will see later how we can fully optimize away XOR bus interactions using just this
+abstraction.
+
+### Memory Bus
+
+TODO Continue with the abstraction using Range Constraints.
+
+## Combining Range Constraints
+
+
+## Optimization Steps
+
+The called functions are
+
+```
+optimize_exec_bus
+loop:
+    solver_based_optimization
+    remove_trivial_constraints
+    remove_free_variables
+    remove_disconnected_columns
+    trivial_simplifications
+    optimize_memory
+    LowDegreeBusInteractionOptimizer
+inliner::replace_constrained_witness_columns
+optimize_range_constraints
+trivial_simplifications
+```
+
+in addition, in the solver we have to explain:
+- linearizing
+- boolean extraction
+- solving algebraic constraints
+ - simple equivalence
+ - splitting into multiple constraints
+ - solving itself
+- handling bus interactions
+- quadratic equivalence detection
+- exhaustive search
+- equal zero check
+
+### Constraint System Solver
```

Comment:
Maybe this section could be moved to the top to clarify what the instances of `solver` above mean.